BF+ENH: Add wxwidgets 3.2.6 and py-wxpython 4.2.2, missing pkgconfig dep

moloney opened the pull request:
Improves compat with newer Python (>3.9) and numpy.

Fix error during configure step (even on at least some older versions) by including 'pkgconfig' as a build dep so gtk+ is found.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

moloney committed:
BF+ENH: Add wxwidgets 3.2.6 and py-wxpython 4.2.2

Improves compat with newer Python (>3.9) and numpy.

Fix error during configure step (even on at least some older
versions) by including 'pkgconfig' as a build dep so gtk+ is found.

wdconinc commented:
I was looking into this yesterday as well and the reason for pkgconfig is that it recompiles wxwidgets instead of reusing it as a dependency; it's wxwidgets that needs pkgconfig.

moloney committed:
BF: Make wxpython use spack built wxwidgets instead of rebuilding

moloney commented:
Good catch, I just pushed a commit that should fix that.

wdconinc commented on var/spack/repos/builtin/packages/py-wxpython/package.py:
```diff
@@ -11,13 +11,15 @@ class PyWxpython(PythonPackage):
     depends_on("cxx", type="build")  # generated
 
     depends_on("wxwidgets +gui")
+    depends_on("wxwidgets@3.2.6 +gui", when="@4.2.2")
```
We try to keep the dependencies correct, even if not all versions are included. Is it possible this should be:
```suggestion
    depends_on("wxwidgets@3.2.6: +gui", when="@4.2:")
```
i.e. all versions 4.2 or higher (including 4.2.2) must have at least wxwidgets version 3.2.6 or higher? 

Or is wxpython only making a statement about whatever version is bundled, and not including any required minimum version? In that case, I'd use something like
```suggestion
    depends_on("wxwidgets@3.2.6:3.2 +gui", when="@4.2.2")
```
since bug fixes in semver should not affect the interface (possibly even `wxwidgets@3.2.6:3` if minor releases are not expected to break compatibility).

moloney replied:
I'm not sure what the minimum version of `wxwidgets` is for any given version of `wxpython` and can't find any documentation that states this.  I just setup this rule since `wxpython@4.2.2` was bundled with `wxwidgets@3.2.6`.  I can potentially try loosening this up, or at least going off semver rules as you suggest.

wdconinc replied:
If this is the  bundled version and there is no additional information, then let's keep as is and hope that there are no conflicts. If there are, we can relax after more study.

wdconinc commented on var/spack/repos/builtin/packages/py-wxpython/package.py:
```diff
@@ -27,3 +29,8 @@ class PyWxpython(PythonPackage):
+    def setup_build_environment(self, env):
+        # By default wxWdigets is built as well instead of using spack provided version,
+        # this tells it to just build the python extensions
+        env.set("WXPYTHON_BUILD_ARGS", "build_py --use_syswx")
```
I wonder if this option existed at the time the first version in spack was added (since it seems someone, *cough* @adamjstewart *cough*, was able to build wxpython at that time without the pkgconfig and gtkplus dependencies). Did you check that providing the `--use_syswx` flag does not lead to failures building the 4.0.6 version if it did not suppport that flag?

moloney replied:
I can give try to test this out soon.

moloney replied:
The 4.0.6 version does understand the `--use_syswx` option, so I assume it was using the bundled wxwidgets, and the build was succeeding somehow with out `pkg_config` included as a build dep.

wdconinc commented on var/spack/repos/builtin/packages/py-wxpython/package.py:
```diff
@@ -11,13 +11,15 @@ class PyWxpython(PythonPackage):
+    depends_on("wxwidgets@3.2.6 +gui", when="@4.2.2")
 
     # Needed for the build.py script
     depends_on("py-setuptools", type="build")
```
While we're applying edits here, I have locally:
```suggestion
    depends_on("py-setuptools", type="build")
    depends_on("py-setuptools@:75", type="build", when="@:4.1")  # deprecated license-file
```
since the next setuptools version will fail to parse the deprecated `license-file` (dashes) field in setup.cfg (ref: https://github.com/wxWidgets/Phoenix/commit/87b606aa041eba8274d6ca50da015f92070d74c6).

wdconinc reviewed (commented):
Checksums verified. Thanks for digging into this and figuring out how to reuse the wxwidgets install (I tried for a bit the other day but then gave up...).

moloney committed:
Update var/spack/repos/builtin/packages/py-wxpython/package.py

Avoid using too new version of py-setuptools as license file format is currently not compatible.

Co-authored-by: Wouter Deconinck <wdconinc@gmail.com>

wdconinc reviewed (approved):
Thanks a lot for this!

wdconinc commented:
Turns out the pkgconfig dependency is still needed even when using external wxwidgets, so apologies for that incorrect comment here. Without pkgconfig, the gdk include directories are not found. See #48565 where I made that chance to get the CI stack to build. Previously failed in e.g. https://gitlab.spack.io/spack/spack/-/jobs/14786523. Upside is that that #48565 will ensure the wxpython build is exercised in CI.
